moped-etl/arcgis/components_to_agol.py

```diff
@@ -100,28 +110,112 @@ def main():
+            logger.info(f"Processing {feature_type} features...")
+            features = all_features[feature_type]
+
+            logger.info("Deleting all existing features...")
+            delete_all_features(feature_type)
+
+            logger.info(
+                f"Uploading {len(features)} features in chunks of {UPLOAD_CHUNK_SIZE}..."
+            )
+            for feature_chunk in chunks(features, UPLOAD_CHUNK_SIZE):
+                logger.info("Uploading chunk....")
+                add_features(feature_type, feature_chunk)
+    else:
+        logger.info("Getting component features from recently updated projects...")
+        data = make_hasura_request(
+            query=COMPONENTS_QUERY_BY_LAST_UPDATE_DATE,
+            variables={"where": {"project_updated_at": {"_gt": args.date}}},
+        )["component_arcgis_online_view"]
+
+        all_features = make_all_features(data)
+
+        # Get unique project IDs that need to have features deleted & replaced
+        project_ids = []
+
+        for component in data:
+            project_ids.append(component["project_id"])
+
+        project_ids_for_feature_delete = list(set(project_ids))
+
+        # Delete outdated feature from AGOL and add updated features
+        for feature_type in ["points", "lines", "combined"]:
+            logger.info(f"Processing {feature_type} features...")
+            logger.info(
+                f"Deleting all existing features in {feature_type} layer for updated projects..."
+            )
+            for delete_chunk in chunks(
+                project_ids_for_feature_delete, UPLOAD_CHUNK_SIZE
+            ):
+                joined_project_ids = ", ".join(str(x) for x in delete_chunk)
+                logger.info(
+                    f"Deleting {len(project_ids_for_feature_delete)} features with project ids {joined_project_ids} in chunks of {UPLOAD_CHUNK_SIZE}..."
+                )
+                delete_features_by_project_ids(feature_type, delete_chunk)
+
+            features = all_features[feature_type]
+
+            logger.info(
+                f"Uploading {len(features)} features in chunks of {UPLOAD_CHUNK_SIZE}..."
+            )
+            for feature_chunk in chunks(features, UPLOAD_CHUNK_SIZE):
+                logger.info("Uploading chunk....")
+                add_features(feature_type, feature_chunk)
 
 
 if __name__ == "__main__":
-    logger = get_logger(__file__)
-    main()
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "-d",
+        "--date",
+        type=str,
+        default=None,
+        help=f"ISO date string of latest updated_at value to find project records to update.",
+    )
+
+    parser.add_argument(
+        "-f",
+        "--full",
+        action="store_true",
+        help=f"Delete and replace all project components.",
+    )
+
+    args = parser.parse_args()
+    logger = get_logger(name="components-to-agol", level=logging.INFO)
+
```

Comment:
it might be good to throw an error here if someone decides to use both `--full` and `--date` in combination: it looks like that would result in the script running both tasks, which wouldn't be a big deal but still 🤷

Reply:
good catch - i'll update 🙏